### PR TITLE
Monei json api

### PR DESCRIPTION
**Pull request by leamarty**
This PR moves from our old (unused) XML API to our new one in JSON, for [MONEI gateway](https://monei.net/)

Biggest difference from before: we are now requesting "API KEY" as credential.
The API core is the same, you'll find almost no changes in the remote tests.
Unit tests were now replaced with JSON responses.

Also, we replaced the API URL so it matches the "homepage_url".

Unit test result:
```
> rake test TEST=test/unit/gateways/monei_test.rb
Loaded suite /Library/Ruby/Gems/2.6.0/gems/rake-13.0.1/lib/rake/rake_test_loader
Started
..............
Finished in 0.139572 seconds.
---------------------------------------------------------------------------------------------------------------------------------------------
14 tests, 56 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed
---------------------------------------------------------------------------------------------------------------------------------------------
100.31 tests/s, 401.23 assertions/s
```

Remote test result:
```
> rake test TEST=test/remote/gateways/remote_monei_test.rb
Loaded suite /Library/Ruby/Gems/2.6.0/gems/rake-13.0.1/lib/rake/rake_test_loader
Started
..................
Finished in 27.320396 seconds.
---------------------------------------------------------------------------------------------------------------------------------------------
18 tests, 42 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed
---------------------------------------------------------------------------------------------------------------------------------------------
0.38 tests/s, 0.89 assertions/s
```

Please let me know if you need anything else to move forward. We had a deadline for a customer that needed to begin operating today... so I hope you understand <3

### CODE REVIEW

File: lib/active_merchant/billing/gateways/monei.rb [hunk 1]
```diff
@@ -245,92 +220,126 @@ def eci_to_result_indicator(eci)
...
 
       # Private: Translate AM operations to Monei operations codes
       def tanslate_payment_code(action)
         {
-          purchase: 'CC.DB',
-          authorize: 'CC.PA',
-          capture: 'CC.CP',
-          refund: 'CC.RF',
-          void: 'CC.RV'
+          purchase: 'sale',
```
Comment (olleolleolle) [inline lib/active_merchant/billing/gateways/monei.rb:327]: This private method name has a typo in it. `tanslate_payment_code` => `translate_payment_code`.

Perhaps this breaking-changes PR is the time to clean that up?

File: lib/active_merchant/billing/gateways/monei.rb [hunk 2]
```diff
@@ -245,92 +220,126 @@ def eci_to_result_indicator(eci)
...
+          authorize: 'auth',
+          capture: 'capture',
+          refund: 'refund',
+          void: 'void'
+        }[action]
+      end
+
+      # Private: Translate AM operations to Monei endpoints
+      def tanslate_action_endpoint(action)
```
Comment (olleolleolle) [inline lib/active_merchant/billing/gateways/monei.rb:336]: Ah, this private method _also_ has the same typo. `tanslate_action_endpoint` => `translate_action_endpoint`

File: lib/active_merchant/billing/gateways/monei.rb (hunk 1 above)
Comment (leamarty) [reply lib/active_merchant/billing/gateways/monei.rb:327]: Oh nice catch! I'll do it, thanks

File: lib/active_merchant/billing/gateways/monei.rb (hunk 2 above)
Comment (leamarty) [reply lib/active_merchant/billing/gateways/monei.rb:336]: Doing

**Comment by leamarty** [comment]
@olleolleolle typo fixed

**Comment by leamarty** [comment]
@olleolleolle great, thank you! Any idea how long it will be until these changes get in Shopify?

**Comment by olleolleolle** [comment]
@leamarty I have no idea. I've no association with Shopify, ActiveMerchant, or anything. I just took a look. Hope this helps!

**Comment by pi3r** [comment]
Hey @leamarty, how do you propose we migrate merchants that are using the previous version of the gateway with different credentials? This is a breaking change and will require that any ActiveMerchant/Monei user changes the passed in credentials.

**Comment by leamarty** [comment]
Hi @pi3r , thanks for reaching out. The good part is that we don't need to care about existing users, because we already took care of that.
We migrated all of them to a hosted payment (redirect) solution before doing this change, so we are super safe here that nobody is using this code by now.

**Comment by pi3r** [comment]
@leamarty We're still seeing some merchants activating Monei on their shop so that doesn't seem 100% accurate. We can keep talking about this subject on the email thread.

**Comment by leamarty** [comment]
Added support for 3DS v2.1. Waiting for a merge

**Comment by alexandresaiz** [comment]
@pi3r hey pierre! Can you take a look? HC from S is waiting for this.

**Review by tatsianaclifton — CHANGES_REQUESTED** [review]
Hello @leamarty and thank you for submitting updates to this gateway.

I've reviewed this submission and have some preliminary questions and requests:

1. Per Active merchant contribution guide it is required to implement 2 non-transactional methods: `scrub(transcript)` and `supports_scrubbing?`. Please add these methods and supporting unit and remote tests.
2. I am curious to know if there are any documents that detail this API for 3ds transactions? I see many not standard options in option hash such as `threed_dynamic`, `sca_exemption`, `user_agent` and others
3. Do I understand correctly that intention of this implementation is only to support credit card transactions?
4. I would like to make sure that other methods such as `verify`, `credit` and `store` are not supported by Monei gateway? I don’t see implementations for them, so just want to double check.

Thanks

**Comment by leamarty** [comment]
Hi @tatsianaclifton thanks for your feedback.
Let me check those points and get back to you asap

**Comment by jimmyn** [comment]
Hi @tatsianaclifton, I have updated the pull request.

1.  `scrub(transcript)` and `supports_scrubbing?` are added along with tests
2. I have removed all non-standard parameters as we are not using them. The only one left is the optional `user_agent` that we are parsing to add customer device information. 
3. As for now we only support credit card transactions
4. `verify` was implemented before, `store` is implemented, integration tests are updated, `credit` method is not supported

Test output:

```
➜  active_merchant git:(monei_json_api) ✗ rake test TEST=test/unit/gateways/monei_test.rb
Loaded suite /Users/dmitriy/.rvm/gems/ruby-2.7.2/gems/rake-13.0.3/lib/rake/rake_test_loader
Started
.................
Finished in 0.100522 seconds.
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
17 tests, 61 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
169.12 tests/s, 606.83 assertions/s
```

```
➜  active_merchant git:(monei_json_api) ✗ rake test TEST=test/remote/gateways/remote_monei_test.rb
Loaded suite /Users/dmitriy/.rvm/gems/ruby-2.7.2/gems/rake-13.0.3/lib/rake/rake_test_loader
Started
.
Finished in 56.774722 seconds.
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
23 tests, 59 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
0.41 tests/s, 1.04 assertions/s
```


Best regards

**Review by tatsianaclifton — CHANGES_REQUESTED** [review]
Thank you for your patience while we are reviewing this PR. I just left couple small comments. Once fixed please follow instructions here to rebase with master and squash commits https://github.com/activemerchant/active_merchant/wiki/Submitting-Your-Gateway#rebasing-after-review

**Comment by jimmyn** [comment]
Hi @tatsianaclifton, I have fixed the issues and squashed the commits. The PR should be ready to merge.

**Comment by tatsianaclifton** [comment]
> Hi @tatsianaclifton, I have fixed the issues and squashed the commits. The PR should be ready to merge.

Hi @jimmyn Code looks good, but I still see multiple commits. We would prefer the merge be done with a single commit, so please squash the remaining commits into one. Thanks

**Comment by jimmyn** [comment]
@tatsianaclifton all done

**Review by tatsianaclifton — COMMENTED** [review]
Looks good

**Comment by bbraschi** [comment]
Hello @leamarty @olleolleolle @tatsianaclifton !
Is this PR no breaking existing users of the active_merchant gem, because new credentials need to be provisioned?
Should such a change not be released through a new Gateway class with its own name?

**Comment by tatsianaclifton** [comment]
> Hello @leamarty @olleolleolle @tatsianaclifton !
> Is this PR no breaking existing users of the active_merchant gem, because new credentials need to be provisioned?
> Should such a change not be released through a new Gateway class with its own name?

Hello @bbraschi That is a good point. You are correct that a new Gateway class is the standard way, but based on this comment https://github.com/activemerchant/active_merchant/pull/3613#issuecomment-627949432 there was no need for a new class as nobody is using the Monei gateway. The PR was submitted by the Monei Gateway team and they have verified that, but maybe something was missed.  Do you know any cases when the old XML API is used and this PR causes any issues? Please let me know asap that I can bring it to my team here at Spreedly for discussion.

**Comment by bbraschi** [comment]
Checking if we now have transactions using old credentials ...

**Comment by bbraschi** [comment]
Hello @jimmyn! It seems that Monei has a restriction on the character set for `orderId` that causes transactions from Shopify to fail. The IDs from Shopify includes a `.`, eg

```
\"orderId\" with value \"#1628035738.1\" fails to match the required pattern: /^[a-zA-Z0-9]{1,40}$/
```        

Can Monei allow `.` for order IDs? TY!

**Comment by jimmyn** [comment]
@bbraschi true, we only support alphanumeric characters in order ID. Are those Shopify order IDs specific to active merchant somehow? We have a redirect flow payment gateway integration with Shopify and never had this problem before. Ids are always integers.

**Comment by bbraschi** [comment]
@jimmym, yes, Shopify order IDs include a `.` with active_merchant for all the active_merchant gateways.

**Comment by jimmyn** [comment]
@bbraschi I need to check what we can do. Can we sanitize this value in active merchant class?

**Comment by bbraschi** [comment]
Yes, it could be done in `MoneiGateway`, but it is possibly going to confuse merchants if the order id in the receipt from Monei does not match Shopify's order ID.

**Pull request closed without merging**